addons/sourcemod/scripting/ConnectAnnounce.sp

```diff
@@ -422,75 +453,121 @@ public Action Command_Announce(int client, int args)
 // #     # #    #  #
 //  #####   #### # #######
 
-stock void SQLInitialize()
+stock bool DB_Connect()
 {
-	if (g_hDatabase != null)
-		delete g_hDatabase;
+	//PrintToServer("DB_Connect(handle %d, state %d, lock %d)", g_hDatabase, g_DatabaseState, g_iConnectLock);
 
-	if (SQL_CheckConfig(DATABASE_NAME))
-		SQL_TConnect(OnSQLConnected, DATABASE_NAME);
-	else
-		SetFailState("Could not find \"%s\" entry in databases.cfg.", DATABASE_NAME);
+	if (g_hDatabase)
+		return true;
+
+	// 100k connections in a minute is bad idea..
+	if (g_DatabaseState == DatabaseState_Wait)
+		return false;
+
+	if (g_DatabaseState != DatabaseState_Connecting)
+	{
+		g_DatabaseState = DatabaseState_Connecting;
+		g_iConnectLock = ++g_iSequence;
+		Database.Connect(GotDatabase, DATABASE_NAME, g_iConnectLock);
+	}
+
+	return false;
 }
 
-stock void OnSQLConnected(Handle hParent, Handle hChild, const char[] err, any data)
+public void GotDatabase(Database db, const char[] error, any data)
 {
-	if (hChild == null)
+	//PrintToServer("GotDatabase(data: %d, lock: %d, g_h: %d, db: %d)", data, g_iConnectLock, g_hDatabase, db);
+
+	// If this happens to be an old connection request, ignore it.
+	if (data != g_iConnectLock || g_hDatabase)
 	{
-		SetFailState("Failed to connect to database \"%s\". (%s)", DATABASE_NAME, err);
+		if (db)
+			delete db;
 		return;
 	}
 
 	char sDriver[16];
-	g_hDatabase = CloneHandle(hChild);
-	SQL_GetDriverIdent(hParent, sDriver, sizeof(sDriver));
-
-	SQL_LockDatabase(g_hDatabase);
+	SQL_GetDriverIdent(g_hDatabase, sDriver, sizeof(sDriver));
 
 	if (!strncmp(sDriver, "my", 2, false))
 		g_bSQLite = false;
 	else
 		g_bSQLite = true;
 
-	SQLSetNames(INVALID_HANDLE);
+	g_iConnectLock = 0;
+	g_DatabaseState = DatabaseState_Connected;
+	g_hDatabase = db;
 
-	SQLTableCreation_Join(INVALID_HANDLE);
+	// See if the connection is valid.
+	if (g_hDatabase)
+		LogMessage("Connected to database.");
+	else
+	{
+		LogError("Connecting to database failed: %s", error);
+		return;
+	}
 
-	SQL_UnlockDatabase(g_hDatabase);
-}
+	// Set character set to UTF8MB4 in the database
+	char sQuery[MAX_SQL_QUERY_LENGTH];
```

Comment:
keep the next 2 queries inside separate functions so that we keep the function GotDatabase as light and understandble as possible